### DISCUSSION TITLE
[entoas] - Query Filter

### DISCUSSION THREAD

**Discussion by masseelch** [ent/ent#2147] [Ideas]
Good Morning everyone,

yesterday I had a user request to add filtering query parameters to the generated spec. There are several possibilities how to tackle this and in order to decide which way to go (first, we can always extend for different strategies later), I opened this discussion.

---
#### Proposal 1 - keep it simple
The most basic solution is to simply map the query parameters to the entity fields and check for equality:

```shell
GET /users?name=ariel&age=10
```
This abviously is a very limited attempt and I don't know id this adds any real value.

#### Proposal 2 - simpler but more useful
Add a special query parameter `terms` (or similar) that are used to search over all the entities fields (blankspace used to `or` values). Constrain further by matching any given entity field on equality. Arrays are used to `or` the values on the field.

```shell
GET /products?terms=hand+bag&productType[]=bag&productType[]=leather&brand=Gucci
```
This solution is more powerful but still relativley simple to implement (for spec implementing servers). One thing to decide though which aray syntax to use. I have seen both `[]array=value1&[]array=value2` and `array=value1&array=value2`

#### Proposal 3 - full blown filter
Have one special query param like `filter` and add a json-encoded string describing the constraints. I am not sure if this it out of scope for this package.

#### Proposal 4 - do it like jhipster
TODO

---

**In all the above we can add an `order` param to define result ordering.**

Any feedback is appreciated. Any other proposals too. Thank you very much :-)

**Comment by rotemtam** [comment]
@masseelch as you may have seen we are working on adding a `List` endpoint to `entproto` (https://github.com/ent/contrib/pull/162).
The next logical step there is to add filters as well. 
I think it would be sensible to think about some common logic between the implementations.

**Reply by rickywei** [reply]
any updates?

**Comment by ysaakpr** [comment]
Would be interested in this, We use Jhipster JDL for rest and crud generation, The JHipster also has a layer which generate the oas spec and rest layer with advanced filtering option, that would be a great place to look for some standards in this area.

https://www.jhipster.tech/

**Comment by a8m** [comment]
I think this feature is super useful. I never used JHipster, but if we provide a filtering capability, it should be flexible. Something like [a8m/rql](https://github.com/a8m/rql) (I created it a few years ago), or the generated [GraphQL Filter Inputs](https://entgo.io/blog/2021/07/01/automatic-graphql-filter-generation).

**Comment by rickywei** [comment]
How is this feature going? If it is done, could you tell me how to find an example for query filter and sort? Thanks

**Reply by migueldv90** [reply]
I was planning on using entoas, but it looks like no development has been done in a while, and never reached a major release. 

"Hey! It is not used internally any more. The current status is most likely what it will stay on. All I can offer is reviews." - MasseElch
